airflow/www/views.py

```diff
@@ -5803,21 +5853,42 @@ def duration_f(self):
     @provide_session
     @action_logging
     def action_clear(self, task_instances, session: Session = NEW_SESSION):
-        """Clears the action."""
+        """Clears an arbitrary number of task instances."""
         try:
-            dag_to_tis = collections.defaultdict(list)
-
-            for ti in task_instances:
-                dag = get_airflow_app().dag_bag.get_dag(ti.dag_id)
-                dag_to_tis[dag].append(ti)
+            count = self._clear_task_instances(
+                task_instances=task_instances, session=session, clear_downstream=False
+            )
+            session.commit()
+            flash(f"{count} task instances have been cleared")
+        except Exception as e:
+            flash(f'Failed to clear task instances: "{e}"', "error")
 
-            for dag, task_instances_list in dag_to_tis.items():
-                models.clear_task_instances(task_instances_list, session, dag=dag)
```

Comment:
Using the same method for both actions is beneficial for maintenance. However, I'm curious if this might impact the performance of the clear action. The `clear_task_instances` method is specifically designed to swiftly clear the provided tasks and their corresponding DAG runs. The newly developed methods clear the task instances on a per-(DAG, DAG run) basis.

Reply:
Good point, during dev I tried clearing a bunch of tis and I didn't notice any visible performance regression but that's by no means a benchmark :)
I'll compare both code paths asap

Reply:
There's indeed an extra ORM call involved for each dag_id, because dag.clear fetches TIs to clear with `_get_task_instances` before calling `clear_task_instances`. So for the default clear action (without downstream) it seems redundant as TIs are already available to the action handler.
I guess perf hit is maximized when the user clears a large number of TIs across a large number of DAG or DAG runs as you hinted (leads to query amplification)
I'll restore the "quick path" for cases when only selected TIs need to be cleared

Reply:
[3977962](https://github.com/apache/airflow/pull/34529/commits/3977962dd471dbd4d69be57787caadde32419a1f)